Add Lesson 7 (#12)
EugeniyBurmistrov/resolve: docs/basics/tutorial/lesson-7.md

## **Lesson 7** - Functionality Enhancements

[\[Get the Code for This Lesson\]](https://github.com/reimagined/resolve/tree/master/examples/shopping-list-tutorial/lesson-7)

> This lesson describes additional steps that you can follow to enhance various aspects of the Shopping List application's functionality.

Reviewer: This is quite vague.

Author: Changed to: 
"The lesson describes how to perform the following enhancements to the Shopping List application's functionality:

- Modify the reSolve backend to provide the complete set of CRUD (create, read, update, delete) operations.
- Modify the frontend to support all CRUD operations.
- Add static resources to the frontend."

Reviewer: "perform the following enhancements" to the "application's functionality"? How can you perform enhancements to its functionality?

Author: 'how to make the following enhancements to the Shopping List application'

Reviewer: how to improve the Shopping List application
OR
how to expand the Shopping List application's functionality

### Modify the Write Side

> Define additional events to implement the full set of CRUD (create, read, update, delete) operations.

Reviewer: "Define additional events" which events?

Author: Changed to 'the following'

Reviewer: Define the following events to implement the full set of CRUD (create, read, update, delete) operations:

Author: done

Reviewer: Ok

**common/event_types.js:**

```js
...
export const SHOPPING_LIST_CREATED = 'SHOPPING_LIST_CREATED'
export const SHOPPING_LIST_RENAMED = 'SHOPPING_LIST_RENAMED'
export const SHOPPING_LIST_REMOVED = 'SHOPPING_LIST_REMOVED'
export const SHOPPING_ITEM_REMOVED = 'SHOPPING_ITEM_REMOVED'
```

> Modify the aggregate projection to enable the shopping list deletion.

Reviewer: ```suggestion
Modify the aggregate projection to allow you to delete the shopping list.
```

    }
  },

  removeShoppingList: state => {
    return {
      type: SHOPPING_LIST_REMOVED
    }
  },

  removeShoppingItem: (state, { payload: { id } }) => {
    return {
      type: SHOPPING_ITEM_REMOVED,
      payload: { id }
    }
  }
```

### Modify the Read Side

Modify the ShoppingList View Model projection to handle the new event types.

**common/view-models/shopping_list.projection.js:**

```js
...

  [SHOPPING_LIST_RENAMED]: (state, { payload: { name } }) => ({
    ...state,
    name
  }),

  [SHOPPING_LIST_REMOVED]: () => ({
    removed: true
  }),

  [SHOPPING_ITEM_REMOVED]: (state, { payload: { id } }) => ({
    ...state,
    list: state.list.filter(item => item.id !== id)
  })
```

Modify the ShoppingLists Read Model projection.

**common/read-models/shopping_lists.projection.js:**

```js
...

  [SHOPPING_LIST_REMOVED]: async (store, { aggregateId }) => {
    await store.delete('ShoppingLists', { id: aggregateId })
  },

  [SHOPPING_LIST_RENAMED]: async (
    store,
    { aggregateId, payload: { name } }
  ) => {
    await store.update('ShoppingLists', { id: aggregateId }, { $set: { name } })
  }
```

### Modify the Frontend

#### Use Static Content

> Add the required static content to the application's **static** folder. The example application uses the following static files:

Reviewer: ```suggestion
Add the static content to the application's **static** folder. The example application uses the following static files:
```

> - The **Styles.css** file - Contains styles used by the application's client components.

Reviewer: ```suggestion
- The **Styles.css** file - Contains styles the application's client components use.
```

- The **fontawesome.min.css** file an the **webfonts** folder - The standard [Font Awesome](https://fontawesome.com/) distribution.

> - The **close-button.png** image - An icon that the button used to remove shopping list items displays.

Reviewer: An icon that the button used?

Author: Changed to 'The Remove Shopping List button's icon.'

Reviewer: Ok

#### Update Components

> Modify the ShoppingLists component to provide UI for shopping list deletion.

Reviewer: "provide UI"? a UI? the UI?

Author: Added 'a'

Reviewer: Ok

**client/components/ShoppingLists.js:**

```js
<th className="example-table-action">Action</th>
...

<td className="example-table-action">
  <Button
    onClick={() => {
      this.props.removeShoppingList(id)
    }}
  >
    <i className="far fa-trash-alt" />
  </Button>
</td>

```

```js
const { lists, createShoppingList, removeShoppingList } = this.props
...
<ShoppingLists lists={lists} removeShoppingList={removeShoppingList} />
```

> Modify the ShoppingList component to provide the capability to rename shopping lists.

Reviewer: ```suggestion
Modify the ShoppingList component to allow users to rename shopping lists.
```

**client/containers/ShoppingList.js:**

```js
state = {
  shoppingListName: this.props.data && this.props.data.name
  ...
}

renameShoppingList = () => {
  this.props.renameShoppingList(this.props.aggregateId, {
    name: this.state.shoppingListName
  })
}

onShoppingListNamePressEnter = event => {
  if (event.charCode === 13) {
    event.preventDefault()
    this.renameShoppingList()
  }
}

updateShoppingListName = event => {
  this.setState({
    shoppingListName: event.target.value
  })
}
...

<FormControl
  type="text"
  value={this.state.shoppingListName}
  onChange={this.updateShoppingListName}
  onKeyPress={this.onShoppingListNamePressEnter}
  onBlur={this.renameShoppingList}
/>
```

Add the list item deletion functionality.

**client/containers/ShoppingList.js:**

```js
const {
  ...
  removeShoppingItem
} = this.props

<Image
  className="example-close-button"
  src="/close-button.png"
  onClick={removeShoppingItem.bind(null, aggregateId, {
    id: todo.id
  })}
/>
```

> The code below demonstrates the **Image** component's implementation.

Reviewer: ```suggestion
The code below implements the **Image** component.
```

**client/containers/Image.js:**

```js
import { Image as BootstrapImage } from 'react-bootstrap'
import { connectStaticBasedUrls } from 'resolve-redux'

const Image = connectStaticBasedUrls(['src'])(BootstrapImage)

export default Image
```

#### Link Stylesheets:

Use the React Helmet component to link stylesheets to your application.

**client/containers/Header.js:**

```js
import {Helmet} from "react-helmet"
...

<Helmet>
  {css.map((href, index) => (
    <link rel="stylesheet" href={href} key={index} />
  ))}
  ...
</Helmet>
...
export default connectStaticBasedUrls(['css', 'favicon'])(Header)
```

**client/containers/App.js:**

```js
<Header
  css={['/fontawesome.min.css', '/style.css', ...]}
  ...
/>
```

#### Update the Optimistic Updates Code

> Modify the code related to optimistic UI updates to support the shopping list deletion.

Reviewer: optimistic?

Author: This is a common software design pattern. For example:
https://medium.com/@_erikaybar/optimistic-ui-updates-in-react-9e139ffa2e45
https://www.reddit.com/r/reactjs/comments/3y80w7/what_is_an_optimistic_update/

Reviewer: Ok
